### ISSUE TITLE
Integrate CrossMPT: Cross-attention Message-Passing Transformer for Error Correcting Codes

### ISSUE THREAD

**Issue by selimfirat** [ipc-lab/kaira#37]
# Integrate CrossMPT: Cross-attention Message-Passing Transformer for Error Correcting Codes

## Summary

Integrate the CrossMPT (Cross-attention Message-Passing Transformer) architecture from ["CrossMPT: Cross-attention Message-Passing Transformer for Error Correcting Codes" (ICLR 2025)](https://arxiv.org/abs/2405.01033) into Kaira as a state-of-the-art neural decoder that bridges the gap between transformer architectures and conventional message-passing decoders.

## Background

CrossMPT represents a significant advancement in neural decoding by introducing a novel approach that combines the power of transformer architectures with the operational principles of conventional message-passing decoders. Unlike existing transformer-based decoders that treat all input vectors uniformly, CrossMPT separates and processes variable node (VN) and check node (CN) representations using cross-attention mechanisms.

### Key Innovation: Cross-Attention Message-Passing

CrossMPT introduces several key innovations over existing approaches:

1. **Dual-Stream Architecture**: Separates variable node and check node representations into distinct streams
2. **Cross-Attention Mechanism**: Updates VN and CN embeddings iteratively using masked cross-attention blocks
3. **Structured Attention Masks**: Attention masks determined by the code's parity-check matrix structure
4. **Message-Passing Equivalence**: Shares operational principles with conventional BP decoders while leveraging transformer benefits

### Advantages over Existing Neural Decoders

1. **Superior Performance**: Significantly outperforms existing neural network-based decoders across various code classes
2. **Computational Efficiency**: Reduced memory usage, complexity, inference time, and training time compared to other transformers
3. **Principled Design**: Explicit connection to conventional message-passing principles
4. **Scalability**: Works effectively across different code lengths and types
5. **Interpretability**: Clear separation of variable and check node processing

### Performance Highlights

- **BCH Codes**: Demonstrates superior BER performance compared to BP, Hyp BP, AR BP, and ECCT decoders
- **LDPC Codes**: Outperforms BP decoders (20, 50, and 100 iterations) on WiMAX LDPC codes
- **Efficiency**: Achieves better performance with significantly reduced computational requirements

#### `kaira/models/fec/decoders/crossmpt_decoder.py`
```python
class CrossMPTDecoder(nn.Module):
    """
    Cross-attention Message-Passing Transformer Decoder
    
    Implements the CrossMPT architecture for neural decoding of linear codes
    using cross-attention between variable node and check node representations.
    """
    
    def __init__(self, 
                 parity_check_matrix: torch.Tensor,
                 d_model: int = 128,
                 num_layers: int = 6, 
                 num_heads: int = 8,
                 dropout: float = 0.0):
        """
        Args:
            parity_check_matrix: Binary parity check matrix (m x n)
            d_model: Model dimension for embeddings
            num_layers: Number of cross-attention encoder layers
            num_heads: Number of attention heads
            dropout: Dropout probability
        """
        
    def forward(self, 
                magnitude: torch.Tensor,
                syndrome: torch.Tensor) -> torch.Tensor:
        """
        Args:
            magnitude: Channel magnitude values (batch_size, n)
            syndrome: Computed syndrome (batch_size, m) 
            
        Returns:
            Decoded bits (batch_size, n)
        """
```

#### Cross-Attention Components
- **CrossMPTEncoder**: Multi-layer encoder with alternating VN/CN cross-attention
- **CrossMPTEncoderLayer**: Single layer with cross-attention and feed-forward
- **CrossAttentionBlock**: Multi-head cross-attention with structured masking
- **PositionwiseFeedForward**: Feed-forward network with GELU activation

#### Attention Mask Generation
```python
def build_crossmpt_attention_masks(parity_check_matrix: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
    """
    Build attention masks for variable node and check node cross-attention.
    
    - VN mask: Variable nodes attend to connected check nodes
    - CN mask: Check nodes attend to connected variable nodes
    - Masks determined by parity check matrix structure
    
    Returns:
        Tuple of (VN_mask, CN_mask) for cross-attention masking
    """
```

#### Decoder Interface Compatibility
```python
# Inherit from existing BaseDecoder interface
class CrossMPTDecoder(BaseDecoder):
    def __init__(self, encoder: LDPCCodeEncoder, **kwargs):
        self.parity_check_matrix = encoder.check_matrix
        # Initialize CrossMPT architecture
        
    def decode(self, received_soft: torch.Tensor) -> torch.Tensor:
        # Compute magnitude and syndrome
        magnitude = torch.abs(received_soft)
        syndrome = self.compute_syndrome(received_soft)
        # Apply CrossMPT decoding
        return self.forward(magnitude, syndrome)
```

#### Training Integration
```python
# Support for various code types already in Kaira
def train_crossmpt_decoder(code_type: str, 
                          code_params: Dict,
                          training_config: Dict):
    """
    Train CrossMPT decoder for specified code type.
    
    Supports:
    - LDPC codes (from RPTU database or custom)
    - BCH codes
    - Polar codes  
    - Custom linear codes
    """
```

#### LDPC Codes (Priority 1)
- Integrate with existing `LDPCCodeEncoder`
- Support RPTU database codes (WiMAX, WiFi, 5G, etc.)
- Custom LDPC matrices

#### BCH Codes (Priority 2)  
- Classical BCH codes for comparison with algebraic decoding
- Support various field sizes and generator polynomials

#### Polar Codes (Priority 3)
- Standard polar code constructions
- Integration with successive cancellation decoding comparison

#### Custom Linear Codes (Priority 4)
- Any linear code with known parity check matrix
- Support for punctured and shortened codes

#### Training Example
```python
# examples/models_fec/train_crossmpt_decoder.py
"""
Train Cross-attention Message-Passing Transformer for LDPC decoding.

Demonstrates:
- CrossMPT decoder setup for WiMAX LDPC codes
- Training with AWGN channel simulation  
- Performance comparison with BP and ECCT decoding
- BER/FER evaluation across SNR range
"""

import torch
from kaira.models.fec.encoders import LDPCCodeEncoder
from kaira.models.fec.decoders import CrossMPTDecoder, BeliefPropagationDecoder
from kaira.channels.analog import AWGNChannel
from kaira.metrics.signal import BitErrorRate, FrameErrorRate

# Create LDPC encoder (WiMAX standard)
encoder = LDPCCodeEncoder(
    rptu_database=True, 
    code_length=648, 
    code_dimension=540, 
    rptu_standart="wimax"
)

# Initialize CrossMPT decoder
crossmpt_decoder = CrossMPTDecoder(
    encoder=encoder,
    d_model=128,
    num_layers=10,
    num_heads=8
)

# Training loop with AWGN channel
for epoch in range(num_epochs):
    for batch in dataloader:
        messages = batch['messages']
        codewords = encoder(messages)
        
        # Add AWGN noise
        channel = AWGNChannel(snr_db=snr)
        received = channel(codewords)
        
        # CrossMPT decoding
        decoded = crossmpt_decoder(received)
        
        # Compute loss and backpropagate
        loss = crossmpt_decoder.loss(decoded, messages, received)
        loss.backward()
        optimizer.step()
```

#### Performance Comparison Example
```python
# examples/models_fec/compare_crossmpt_vs_others.py
"""
Performance comparison: CrossMPT vs BP vs ECCT decoding.

Shows superior performance of CrossMPT while maintaining 
educational value of understanding different approaches.
"""

def compare_decoders():
    # Test CrossMPT, BP, and ECCT decoders
    crossmpt_ber = evaluate_decoder(crossmpt_decoder, test_data)
    bp_ber = evaluate_decoder(bp_decoder, test_data)
    ecct_ber = evaluate_decoder(ecct_decoder, test_data)
    
    # Plot comparison showing CrossMPT advantages
    plot_ber_comparison(crossmpt_ber, bp_ber, ecct_ber)
```

### Phase 1: Core Architecture (3-4 weeks)
- [ ] Implement `CrossMPTDecoder` base class
- [ ] Create cross-attention encoder components
- [ ] Implement variable node and check node embedding layers
- [ ] Add attention mask generation from parity check matrices
- [ ] Develop syndrome computation utilities
- [ ] Unit tests for architecture components

### Phase 2: Training Framework (2-3 weeks)  
- [ ] Integrate with existing Kaira training infrastructure
- [ ] Add CrossMPT-specific loss functions and optimizers
- [ ] Support for different code types (LDPC, BCH, Polar)
- [ ] Training utilities and data loaders
- [ ] Integration tests with existing encoders

### Phase 3: Examples and Documentation (2 weeks)
- [ ] Create comprehensive training examples
- [ ] Performance comparison examples (CrossMPT vs BP vs ECCT)
- [ ] Tutorial notebooks for different code types
- [ ] API documentation and usage guides
- [ ] Benchmark results and performance analysis

### Phase 4: Advanced Features (2-3 weeks)
- [ ] Support for punctured and shortened codes
- [ ] Mixed-precision training optimization
- [ ] Distributed training support
- [ ] Model compression and quantization techniques
- [ ] Advanced attention pattern optimizations

### Cross-Attention Architecture
The key innovation is the dual-stream processing with cross-attention:
```python
def forward(self, x_vn, x_cn, mask_vn, mask_cn):
    """
    Alternating cross-attention between VN and CN representations
    """
    for layer in self.encoder_layers:
        # Variable nodes attend to check nodes
        x_vn = layer.cross_attention(x_vn, x_cn, x_cn, mask_vn)
        # Check nodes attend to variable nodes  
        x_cn = layer.cross_attention(x_cn, x_vn, x_vn, mask_cn)
        
        # Feed-forward processing
        x_vn = layer.feed_forward(x_vn)
        x_cn = layer.feed_forward(x_cn)
    
    return x_vn, x_cn
```

### Attention Mask Construction
Build separate masks for variable node and check node attention:
```python
def build_attention_masks(parity_check_matrix):
    m, n = parity_check_matrix.shape  # m=check nodes, n=variable nodes
    
    # VN mask: Variable nodes attend to connected check nodes
    vn_mask = torch.zeros(n, m)
    for i in range(n):
        connected_checks = torch.where(parity_check_matrix[:, i] > 0)[0]
        vn_mask[i, connected_checks] = 1
    
    # CN mask: Check nodes attend to connected variable nodes
    cn_mask = torch.zeros(m, n)
    for i in range(m):
        connected_vars = torch.where(parity_check_matrix[i, :] > 0)[0]
        cn_mask[i, connected_vars] = 1
        
    # Invert for attention masking (0 = attend, 1 = mask)
    return ~vn_mask.bool(), ~cn_mask.bool()
```

### Model Architecture
```
Input: [magnitude(n), syndrome(m)]
  ↓
Embedding: 
  - VN_embed: magnitude → (n, d_model)
  - CN_embed: syndrome → (m, d_model)
  ↓
CrossMPT Encoder (N layers):
  - VN Cross-Attention (attend to CN)
  - CN Cross-Attention (attend to VN)  
  - Position-wise Feed-Forward
  - Layer Normalization & Residual Connections
  ↓
Output Processing:
  - Concatenate: [VN_final, CN_final] → (n+m, d_model)
  - Linear Projection: (n+m, d_model) → (n+m, 1)
  - Final Linear: (n+m,) → (n,)
  ↓
Decoded Bits
```

### Loss Function
Binary cross-entropy with logits for bit-level decoding:
```python
def crossmpt_loss(logits, target_bits, received_soft):
    # Binary cross-entropy loss
    bce_loss = F.binary_cross_entropy_with_logits(logits, target_bits)
    
    # Optional: Add syndrome consistency constraint
    decoded_bits = torch.sigmoid(logits)
    syndrome_loss = syndrome_consistency_penalty(decoded_bits, parity_matrix)
    
    return bce_loss + syndrome_loss
```

### Research Benefits
1. **State-of-the-Art Performance**: Access to ICLR 2025 cutting-edge neural decoding
2. **Message-Passing Connection**: Bridge between neural and conventional approaches
3. **Flexible Experimentation**: Easy to modify cross-attention patterns for research
4. **Comparative Studies**: Direct comparison with BP, ECCT, and other methods
5. **Code Agnostic**: Works with any linear code type

### Educational Benefits  
1. **Cross-Attention Understanding**: Learn transformer cross-attention in ECC context
2. **Message-Passing Insights**: See connection between neural and conventional decoders
3. **Attention Visualization**: Understand how attention patterns reflect code structure
4. **Architecture Comparison**: Compare self-attention (ECCT) vs cross-attention (CrossMPT)
5. **Implementation Deep-Dive**: Understand advanced transformer architectures

### Practical Benefits
1. **Superior Performance**: Significant improvement over existing neural methods
2. **Computational Efficiency**: Reduced memory, complexity, and training time
3. **Production Ready**: Trained models can be exported for deployment
4. **Scalable**: Works across different code lengths and rates
5. **Robust**: Demonstrated performance across multiple code families

## Success Metrics

1. **Performance**: CrossMPT decoder achieves reported performance gains vs BP and ECCT
2. **Compatibility**: Seamless integration with existing Kaira encoders and infrastructure
3. **Usability**: Clear examples and documentation for easy adoption
4. **Efficiency**: Training and inference times competitive with original implementation
5. **Extensibility**: Easy to extend for new code types and research directions
6. **Benchmarks**: Reproduce key results from ICLR 2025 paper

## References

- **Paper**: [CrossMPT: Cross-attention Message-Passing Transformer for Error Correcting Codes](https://arxiv.org/abs/2405.01033) (ICLR 2025)
- **Original Code**: [CrossMPT GitHub Repository](https://github.com/iil-postech/crossmpt)
- **Authors**: Seong-Joon Park, Hee-Youl Kwak, Sang-Hyo Kim, Yongjune Kim, Jong-Seon No
- **Affiliation**: POSTECH Information and Intelligence Lab (IIL)

## Implementation Notes

- Maintain compatibility with existing Kaira decoder interface
- Leverage existing infrastructure for channel simulation and metrics
- Ensure proper documentation and examples for educational use
- Consider computational requirements and provide optimization guidance
- Support both research and production use cases
- Build upon ECCT integration work for consistency

This integration will significantly enhance Kaira's neural decoding capabilities by adding the most recent state-of-the-art transformer architecture while maintaining the library's focus on education, research, and practical applications. The cross-attention mechanism provides a principled connection to conventional message-passing decoders, making it an ideal addition to Kaira's educational mission.
